### PR TITLE
server: deregister should be call if stream revc error after inject net error

### LINKED ISSUE #11341: snapshot count may not degister after inject network fault
### What version of TiKV are you using?
/ # ./tikv-server -V
TiKV
Release Version: 5.3.0-alpha
Edition: Community
Git Commit Hash: d628f0aa5ff8f6ac258ec9989f6886a1dba34fe4
Git Commit Branch: heads/refs/tags/v5.3.0
UTC Build Time: 2021-11-10 14:30:28
Rust Version: rustc 1.56.0-nightly (2faabf579 2021-07-27)
Enable Features: jemalloc mem-profiling portable sse protobuf-codec test-engines-rocksdb cloud-aws cloud-gcp
Profile: dist_release

### Steps to reproduce
1、inject network fault
2、scale out a new tikv（tikv-4）

### What did you expect?
the receive snapshot should be zero if the net work recover 
### What did happened?
pd receive tikv-4 report：
```
"status": {
       "capacity": "1.433TiB",
       "available": "1.353TiB",
       "used_size": "6.874GiB",
       "leader_count": 206,
       "leader_weight": 1,
       "leader_score": 206,
       "leader_size": 17422,
       "region_count": 206,
       "region_weight": 1,
       "region_score": 17422,
       "region_size": 17422,
       "slow_score": 1,
       "receiving_snap_count": 71,
       "start_ts": "2021-11-10T15:55:13Z",
       "last_heartbeat_ts": "2021-11-11T06:59:18.691694279Z",
       "uptime": "15h4m5.691694279s"
     }
```
and some metircs in grafana
image

tikv-0 log：
`[2021/11/10 16:24:25.664 +00:00] [ERROR] [snap.rs:456] ["failed to send snap"] [err=Grpc(RemoteStopped)] [to_addr=tc-tikv-4.tc-tikv-peer.endless-oltp--tps-360624-1-365.svc:20160]`

### PR DESCRIPTION
**Pull request by bufferflies**
Signed-off-by: bufferflies <1045931706@qq.com>

### What problem does this PR solve?
Issue Number: close #11341

Problem Summary:

### What is changed and how it works?
deregister should run even if receiving error
Proposal: [xxx](url) 

### Check List 

Tests 

Side effects

### Release note 

```release-note
None
```

### CODE REVIEW

**Comment by nolouch** [comment]
Which PR introduced this?

**Comment by bufferflies** [comment]
> Which PR introduced this?

refer #11341

**Review by lhy1024 — APPROVED** [review]

**Review by nolouch — APPROVED** [review]

**Review by JmPotato — APPROVED** [review]

**Review by hicqu — APPROVED** [review]

**Comment by BusyJay** [comment]
This PR introduces an unstable case test_sending_fail_with_net_error.

File: tests/failpoints/cases/test_snap.rs [hunk 1]
```diff
@@ -540,3 +540,29 @@ fn test_snapshot_gc_after_failed() {
...
+    // store2 will not receive any data from stream
+    fail::cfg(on_send_store_fp, "return()").unwrap();
+    pd_client.add_peer(1, new_learner_peer(2, 2));
+
+    let engine2 = cluster.get_engine(2);
+    must_get_none(&engine2, b"k1");
+
+    // wait for sending snapshot
+    sleep_ms(100);
+    assert_eq!(cluster.get_snap_mgr(1).stats().sending_count, 0);
```
Comment (BusyJay) [inline tests/failpoints/cases/test_snap.rs:566]: This is racy. Because snapshot will always be retried, you can't know whether it will be 0 at this time.

**Pull request merged**